File: platforms/ios/Packages/FunputKit/Sources/FunputShared/Shortcuts/ShortcutLibrary.swift
import Foundation

/// Versioned, standalone App Group document; not part of keyboard bootstrap settings.
public struct ShortcutLibrary: Codable, Hashable, Sendable {
    public static let currentSchemaVersion = 1
    public var schemaVersion: Int
    public var entries: [TextShortcut]
    public var isEnabled: Bool
    public var smartCase: Bool
    public var inEnglish: Bool

    public init(
        entries: [TextShortcut] = [],
        isEnabled: Bool = true,
        smartCase: Bool = true,
        inEnglish: Bool = true,
        schemaVersion: Int = Self.currentSchemaVersion
    ) {
        self.schemaVersion = schemaVersion
        self.entries = entries
        self.isEnabled = isEnabled
        self.smartCase = smartCase
        self.inEnglish = inEnglish
    }

    public func isDuplicate(_ entry: TextShortcut) -> Bool {
        entries.contains { $0.id != entry.id && $0.trigger == entry.trigger }
    }

    /// Also enforced on disk reads, so malformed data cannot silently become editable.
    public func validate() throws {
        guard schemaVersion == Self.currentSchemaVersion else {
            throw ShortcutsStorageError.unsupportedVersion
        }
        var ids = Set<UUID>()
        var triggers = Set<String>()
        for entry in entries {
            guard entry.isValid, ids.insert(entry.id).inserted else {
                throw ShortcutsStorageError.invalidData
            }
            guard triggers.insert(entry.trigger).inserted else {
                throw ShortcutsStorageError.duplicateTrigger
            }
        }
    }
}

File: platforms/android/shortcut-store/src/main/java/app/funput/funput/shortcuts/model/ShortcutLibrary.kt
package app.funput.funput.shortcuts.model

import app.funput.funput.shortcuts.persistence.ShortcutsStorageError

/** Versioned standalone document shared by the app and input method. */
data class ShortcutLibrary(
    val entries: List<TextShortcut> = emptyList(),
    val isEnabled: Boolean = true,
    val smartCase: Boolean = true,
    val inEnglish: Boolean = true,
    val schemaVersion: Int = CurrentSchemaVersion,
) {
    fun isDuplicate(entry: TextShortcut): Boolean =
        entries.any { it.id != entry.id && it.trigger == entry.trigger }

    fun validate() {
        if (schemaVersion != CurrentSchemaVersion) throw ShortcutsStorageError.UnsupportedVersion
        val ids = HashSet<java.util.UUID>()
        val triggers = HashSet<String>()
        entries.forEach { entry ->
            if (!entry.isValid || !ids.add(entry.id)) throw ShortcutsStorageError.InvalidData
            if (!triggers.add(entry.trigger)) throw ShortcutsStorageError.DuplicateTrigger
        }
    }

    companion object {
        const val CurrentSchemaVersion = 1
    }
}
